File: Episode024/WebStencilsDemos/WebBrokerProject/C++/UtilsFormSession.cpp
//---------------------------------------------------------------------------
#include <vcl.h>
#pragma hdrstop

#include "UtilsFormSession.h"
#include "UtilsLogger.h"
#include <System.DateUtils.hpp>

//---------------------------------------------------------------------------
#pragma package(smart_init)

// Static member definitions
const String TFormSessionManager::FORM_DATA_PREFIX = "FormData_";
const String TFormSessionManager::FORM_ERRORS_PREFIX = "FormErrors_";
const String TFormSessionManager::FORM_TIMESTAMP_SUFFIX = "_timestamp";

//---------------------------------------------------------------------------
String TFormSessionManager::GetFormDataKey(const String& AFormName, const String& AFieldName)
{
    return FORM_DATA_PREFIX + AFormName + "_" + AFieldName;
}

//---------------------------------------------------------------------------
String TFormSessionManager::GetFormErrorsKey(const String& AFormName)
{
    return FORM_ERRORS_PREFIX + AFormName;
}

//---------------------------------------------------------------------------
String TFormSessionManager::GetFormTimestampKey(const String& AFormName)
{
    return FORM_DATA_PREFIX + AFormName + FORM_TIMESTAMP_SUFFIX;
}

//---------------------------------------------------------------------------
void TFormSessionManager::StoreFormData(TWebSession* ASession, const String& AFormName, TWebRequest* ARequest)
{
    if (!ASession || !ARequest || AFormName.IsEmpty())
    {
        Logger->Warning("StoreFormData: Invalid parameters");
        return;
    }
    
    try
    {
        // Store each form field in session
        for (int I = 0; I < ARequest->ContentFields->Count; I++)
        {
            String FieldName = ARequest->ContentFields->Names[I];
            String FieldValue = ARequest->ContentFields->ValueFromIndex[I];
            String SessionKey = GetFormDataKey(AFormName, FieldName);
            // Use AddPair to mimic Delphi behavior and ensure persistence
            ASession->DataVars->AddPair(SessionKey, FieldValue);
        }
        
        // Set timestamp for cleanup
        String TimestampKey = GetFormTimestampKey(AFormName);
        ASession->DataVars->AddPair(TimestampKey, FormatDateTime("yyyy-mm-dd hh:nn:ss", Now()));
        
    }
    catch (Exception &E)
    {
        Logger->Error(String("Error storing form data: ") + E.Message);
    }
}

//---------------------------------------------------------------------------
void TFormSessionManager::StoreValidationErrors(TWebSession* ASession, const String& AFormName, const std::vector<String>& AErrors)
{
    if (!ASession || AFormName.IsEmpty())
    {
        Logger->Warning("StoreValidationErrors: Invalid parameters");
        return;
    }
    
    try
    {
        String ErrorsKey = GetFormErrorsKey(AFormName);
        
        // Convert vector to delimited string
        String ErrorsString = "";
        for (size_t I = 0; I < AErrors.size(); I++)
        {
            if (I > 0)
                ErrorsString += "|";
            ErrorsString += AErrors[I];
        }
        
        ASession->DataVars->Values[ErrorsKey] = ErrorsString;
        
    }
    catch (Exception &E)
    {
        Logger->Error(String("Error storing validation errors: ") + E.Message);
    }
}

//---------------------------------------------------------------------------
std::unique_ptr<TStringDictionary> TFormSessionManager::GetFormData(TWebSession* ASession, const String& AFormName)
{
    auto FormData = std::make_unique<TStringDictionary>();
    
    if (!ASession || AFormName.IsEmpty())
    {
        Logger->Warning("GetFormData: Invalid parameters");
        return FormData;
    }
    
    try
    {
        String KeyPrefix = FORM_DATA_PREFIX + AFormName + "_";
        
        // Iterate through session data variables
        for (int I = 0; I < ASession->DataVars->Count; I++)
        {
            String Key = ASession->DataVars->Names[I];
            if (Key.Pos(KeyPrefix) == 1) // Key starts with prefix
            {
                String FieldName = Key.SubString(KeyPrefix.Length() + 1, Key.Length());
                // Skip timestamp field
                if (FieldName != "timestamp")
                {
                    // Use Values[] or ValueFromIndex as available; prefer Values[]
                    String Value = ASession->DataVars->Values[Key];
                    FormData->Add(FieldName, Value);
                }
            }
        }
        
    }
    catch (Exception &E)
    {
        Logger->Error(String("Error getting form data: ") + E.Message);
        FormData->Clear();
    }
    
    return FormData;
}

//---------------------------------------------------------------------------
std::vector<String> TFormSessionManager::GetValidationErrors(TWebSession* ASession, const String& AFormName)
{
    std::vector<String> Errors;
    
    if (!ASession || AFormName.IsEmpty())
    {
        Logger->Warning("GetValidationErrors: Invalid parameters");
        return Errors;
    }
    
    try
    {
        String ErrorsKey = GetFormErrorsKey(AFormName);
        String ErrorsString = ASession->DataVars->Values[ErrorsKey];
        
        if (!ErrorsString.IsEmpty())
        {
            // Split delimited string into vector
            int Pos = 1;
            while (Pos <= ErrorsString.Length())
            {
                int NextPos = ErrorsString.Pos("|");
                if (NextPos == 0 || NextPos < Pos)
                    NextPos = ErrorsString.Length() + 1;
                    
                String Error = ErrorsString.SubString(Pos, NextPos - Pos);
                if (!Error.IsEmpty())
                    Errors.push_back(Error);
                    
                Pos = NextPos + 1;
            }
        }
        
    }
    catch (Exception &E)
    {
        Logger->Error(String("Error getting validation errors: ") + E.Message);
        Errors.clear();
    }
    
    return Errors;
}

//---------------------------------------------------------------------------
void TFormSessionManager::ClearFormData(TWebSession* ASession, const String& AFormName)
{
    if (!ASession || AFormName.IsEmpty())
    {
        Logger->Warning("ClearFormData: Invalid parameters");
        return;
    }
    
    try
    {
        String KeyPrefix = FORM_DATA_PREFIX + AFormName + "_";
        String ErrorsPrefix = FORM_ERRORS_PREFIX + AFormName + "_";
        String ErrorsKey = GetFormErrorsKey(AFormName);
        
        // Remove form data keys and field-specific error keys
        for (int I = ASession->DataVars->Count - 1; I >= 0; I--)
        {
            String Key = ASession->DataVars->Names[I];
            if (Key.Pos(KeyPrefix) == 1 || Key.Pos(ErrorsPrefix) == 1 || Key == ErrorsKey)
            {
                ASession->DataVars->Delete(I);
            }
        }
        
        Logger->Debug(String("Cleared form data for form: ") + AFormName);
    }
    catch (Exception &E)
    {
        Logger->Error(String("Error clearing form data: ") + E.Message);
    }
}

//---------------------------------------------------------------------------
bool TFormSessionManager::HasFormData(TWebSession* ASession, const String& AFormName)
{
    if (!ASession || AFormName.IsEmpty())
    {
        return false;
    }
    
    try
    {
        String TimestampKey = GetFormTimestampKey(AFormName);
        String Timestamp = ASession->DataVars->Values[TimestampKey];
        return !Timestamp.IsEmpty();
    }
    catch (Exception &E)
    {
        Logger->Error(String("Error checking form data: ") + E.Message);
        return false;
    }
}

//---------------------------------------------------------------------------
void TFormSessionManager::StoreFieldErrors(TWebSession* ASession, const String& AFormName, TStringDictionary* AErrors)
{
    if (!ASession || !AErrors)
        return;

    try
    {
        // Persist each field error under a unique key
        DynamicArray<String> Keys = AErrors->Keys->ToArray();
        const int Count = Keys.get_length();
        for (int i = 0; i < Count; ++i)
        {
            const String FieldName = Keys[i];
            const String ErrorMsg = AErrors->Items[FieldName];
            const String Key = FORM_ERRORS_PREFIX + AFormName + "_" + FieldName.UpperCase();
            ASession->DataVars->Values[Key] = ErrorMsg;
        }
        // Also store a count for cleanup
        ASession->DataVars->Values[FORM_ERRORS_PREFIX + AFormName + "_count"] = IntToStr(Count);
    }
    catch (Exception &E)
    {
        Logger->Error(String("Error storing field errors: ") + E.Message);
    }
}

//---------------------------------------------------------------------------
std::unique_ptr<TStringDictionary> TFormSessionManager::GetFieldErrors(TWebSession* ASession, const String& AFormName)
{
    auto Errors = std::make_unique<TStringDictionary>();
    if (!ASession)
        return std::move(Errors);

    try
    {
        const String Prefix = FORM_ERRORS_PREFIX + AFormName + "_";
        for (int i = 0; i < ASession->DataVars->Count; ++i)
        {
            const String Key = ASession->DataVars->Names[i];
            if (Key.Pos(Prefix) == 1)
            {
                // Extract tail after prefix and skip the special "count" key
                const String Tail = Key.SubString(Prefix.Length() + 1, Key.Length());
                if (!Tail.IsEmpty() && Tail != "count")
                {
                    const String FieldName = Tail;
                    const String Value = ASession->DataVars->ValueFromIndex[i];
                    Errors->AddOrSetValue(FieldName, Value);
                }
            }
        }
    }
    catch (Exception &E)
    {
        Logger->Error(String("Error getting field errors: ") + E.Message);
        Errors->Clear();
    }

    return Errors;
}

File: Episode024/WebStencilsDemos/WebBrokerProject/Delphi/Utils.FormSession.pas
unit Utils.FormSession;

interface

uses
  Web.HTTPApp, 
  System.Generics.Collections,
  System.SysUtils,
  Utils.Logger;

type
  TFormSessionManager = class
  private
    const
      FORM_DATA_PREFIX = 'FormData_';
      FORM_ERRORS_PREFIX = 'FormErrors_';
      FORM_TIMESTAMP_SUFFIX = '_timestamp';
  public
    class procedure StoreFormData(ASession: TWebSession; const AFormName: string; ARequest: TWebRequest);
    class procedure StoreValidationErrors(ASession: TWebSession; const AFormName: string; const AErrors: TArray<string>);
    class function GetFormData(ASession: TWebSession; const AFormName: string): TDictionary<string, string>;
    class function GetValidationErrors(ASession: TWebSession; const AFormName: string): TArray<string>;
    class procedure ClearFormData(ASession: TWebSession; const AFormName: string);
    class function HasFormData(ASession: TWebSession; const AFormName: string): Boolean;
  end;

implementation

uses
  System.DateUtils;

{ TFormSessionManager }

class procedure TFormSessionManager.StoreFormData(ASession: TWebSession; const AFormName: string; ARequest: TWebRequest);
var
  I: Integer;
  SessionKey, SessionContent: string;
begin
  if not Assigned(ASession) or not Assigned(ARequest) then
  begin
    Logger.Warning('StoreFormData: Invalid parameters');
    Exit;
  end;
    
  try
    // Store each form field in session
    for I := 0 to ARequest.ContentFields.Count - 1 do
    begin
      SessionKey := FORM_DATA_PREFIX + AFormName + '_' + ARequest.ContentFields.Names[I];
      SessionContent := ARequest.ContentFields.ValueFromIndex[I];
      ASession.DataVars.AddPair(SessionKey, SessionContent);
    end;
    
    // Set timestamp for cleanup
    ASession.DataVars.AddPair(FORM_DATA_PREFIX + AFormName + FORM_TIMESTAMP_SUFFIX, DateTimeToStr(Now));
  except
    on E: Exception do
      Logger.Error(Format('Error storing form data: %s', [E.Message]));
  end;
end;

class procedure TFormSessionManager.StoreValidationErrors(ASession: TWebSession; const AFormName: string; const AErrors: TArray<string>);
var
  I: Integer;
  SessionKey: string;
begin
  if not Assigned(ASession) or (Length(AErrors) = 0) then
  begin
    Logger.Warning('StoreValidationErrors: Invalid parameters');
    Exit;
  end;
    
  try
    // Store each error in session
    for I := 0 to High(AErrors) do
    begin
      SessionKey := FORM_ERRORS_PREFIX + AFormName + '_' + IntToStr(I);
      ASession.DataVars.Values[SessionKey] := AErrors[I];
    end;
    
    // Store error count
    ASession.DataVars.Values[FORM_ERRORS_PREFIX + AFormName + '_count'] := IntToStr(Length(AErrors));
  except
    on E: Exception do
      Logger.Error(Format('Error storing validation errors: %s', [E.Message]));
  end;
end;

class function TFormSessionManager.GetFormData(ASession: TWebSession; const AFormName: string): TDictionary<string, string>;
var
  I: Integer;
  SessionKey, FieldName, FieldValue: string;
  Prefix: string;
begin
  Result := TDictionary<string, string>.Create;
  
  if not Assigned(ASession) then
  begin
    Logger.Warning('GetFormData: Invalid parameters');
    Exit;
  end;
    
  try
    Prefix := FORM_DATA_PREFIX + AFormName + '_';
    
    // Extract form data from session using DataVars
    for I := 0 to ASession.DataVars.Count - 1 do
    begin
      SessionKey := ASession.DataVars.Names[I];
      if SessionKey.StartsWith(Prefix) and not SessionKey.EndsWith(FORM_TIMESTAMP_SUFFIX) then
      begin
        FieldName := SessionKey.Replace(Prefix, '');
        FieldValue := ASession.DataVars.Values[SessionKey];
        Result.Add(FieldName, FieldValue);
      end;
    end;
  except
    on E: Exception do
      Logger.Error(Format('Error getting form data: %s', [E.Message]));
  end;
end;

class function TFormSessionManager.GetValidationErrors(ASession: TWebSession; const AFormName: string): TArray<string>;
var
  I, ErrorCount: Integer;
  SessionKey: string;
  ErrorList: TList<string>;
begin
  SetLength(Result, 0);
  
  if not Assigned(ASession) then
    Exit;
    
  // Get error count
  SessionKey := FORM_ERRORS_PREFIX + AFormName + '_count';
  ErrorCount := StrToIntDef(ASession.DataVars.Values[SessionKey], 0);
  
  if ErrorCount = 0 then
    Exit;
    
  ErrorList := TList<string>.Create;
  try
    // Retrieve all errors
    for I := 0 to ErrorCount - 1 do
    begin
      SessionKey := FORM_ERRORS_PREFIX + AFormName + '_' + IntToStr(I);
      ErrorList.Add(ASession.DataVars.Values[SessionKey]);
    end;
    
    Result := ErrorList.ToArray;
  finally
    ErrorList.Free;
  end;
end;

class procedure TFormSessionManager.ClearFormData(ASession: TWebSession; const AFormName: string);
var
  I: Integer;
  SessionKey: string;
  KeysToRemove: TList<string>;
begin
  if not Assigned(ASession) then
  begin
    Logger.Warning('ClearFormData: Invalid parameters');
    Exit;
  end;
    
  try
    KeysToRemove := TList<string>.Create;
    try
      // Find all keys related to this form
      for I := 0 to ASession.DataVars.Count - 1 do
      begin
        SessionKey := ASession.DataVars.Names[I];
        if SessionKey.StartsWith(FORM_DATA_PREFIX + AFormName + '_') or
           SessionKey.StartsWith(FORM_ERRORS_PREFIX + AFormName + '_') then
        begin
          KeysToRemove.Add(SessionKey);
        end;
      end;
      
      // Remove all found keys by setting them to empty
      for SessionKey in KeysToRemove do
      begin
        I := ASession.DataVars.IndexOfName(SessionKey);
        if I >= 0 then
          ASession.DataVars.Delete(I);
      end;
    finally
      KeysToRemove.Free;
    end;
    
    Logger.Debug(Format('Cleared form data for form: %s', [AFormName]));
  except
    on E: Exception do
      Logger.Error(Format('Error clearing form data: %s', [E.Message]));
  end;
end;

class function TFormSessionManager.HasFormData(ASession: TWebSession; const AFormName: string): Boolean;
var
  I: Integer;
  SessionKey: string;
  Prefix: string;
begin
  Result := False;
  
  if not Assigned(ASession) then
  begin
    Logger.Warning('HasFormData: Invalid parameters');
    Exit;
  end;
    
  try
    Prefix := FORM_DATA_PREFIX + AFormName + '_';
    
    // Check if any form data exists
    for I := 0 to ASession.DataVars.Count - 1 do
    begin
      SessionKey := ASession.DataVars.Names[I];
      if SessionKey.StartsWith(Prefix) and not SessionKey.EndsWith(FORM_TIMESTAMP_SUFFIX) then
      begin
        Result := True;
        Break;
      end;
    end;
  except
    on E: Exception do
      Logger.Error(Format('Error checking form data: %s', [E.Message]));
  end;
end;

end. 